src/Microsoft.AspNetCore.OData/Extensions/Endpoint/ODataEndpointLinkGenerator.cs

```diff
@@ -0,0 +1,135 @@
+            {
+                string odataPath = odataPathValue as string;
+                if (odataPath != null)
+                {
+                    IPerRouteContainer perRouteContainer = httpContext.RequestServices.GetRequiredService<IPerRouteContainer>();
+                    string routePrefix = perRouteContainer.GetRoutePrefix(httpContext.Request.ODataFeature().RouteName);
+
+                    bool canGenerateDirectLink = routePrefix == null || routePrefix.IndexOf('{') == -1;
+                    if (!canGenerateDirectLink)
+                    {
+                        routePrefix = BindPrefixTemplate(routePrefix, values, ambientValues);
+                    }
+                    string link = CombinePathSegments(routePrefix, odataPath);
+                    link = UriEncode(link);
+                    return link;
+                }
+            }
+
+            return _innerGenerator.GetPathByAddress(httpContext, address, values, ambientValues, pathBase, fragment, options);
+        }
+
+        /// <inheritdoc/>
+        public override string GetPathByAddress<TAddress>(TAddress address, RouteValueDictionary values, PathString pathBase = default, FragmentString fragment = default, LinkOptions options = null)
+        {
+            return _innerGenerator.GetPathByAddress<TAddress>(address, values, pathBase, fragment, options);
+        }
+
+        /// <inheritdoc/>
+        public override string GetUriByAddress<TAddress>(HttpContext httpContext, TAddress address, RouteValueDictionary values, RouteValueDictionary ambientValues = null, string scheme = null, HostString? host = null, PathString? pathBase = null, FragmentString fragment = default, LinkOptions options = null)
+        {
+            return _innerGenerator.GetUriByAddress(httpContext, address, values, ambientValues, scheme, host, pathBase, fragment, options);
+        }
+
+        /// <inheritdoc/>
+        public override string GetUriByAddress<TAddress>(TAddress address, RouteValueDictionary values, string scheme, HostString host, PathString pathBase = default, FragmentString fragment = default, LinkOptions options = null)
+        {
+            return _innerGenerator.GetUriByAddress(address, values, scheme, host, pathBase, fragment, options);
+        }
+
+        // Noted: simple workaround to bind the value for the prefix template
+        // Shold replace it using the standard ASP.NET Core way later.
+        private static string BindPrefixTemplate(string prefix, RouteValueDictionary values, RouteValueDictionary ambientValues)
+        {
+            IList<string> templates = new List<string>();
+            int startIndex = 0;
+            while (true)
+            {
+                int start = prefix.IndexOf('{', startIndex);
+                int end = prefix.IndexOf('}', start + 1); // start == -1 is ok
+                if (start == -1 || end == -1)
+                {
+                    break;
+                }
```

Comment:
I am missing something may be. It seems that the if prefix has '{' and '}', this will loop indefinitely as prefix never changes.

Reply:
No, see my test cases for more information.

---
In reply to: [376002351](https://github.com/OData/WebApi/pull/2035#discussion_r376002351) [](ancestors = 376002351)